docs/architecture/adr-028-public-key-addresses.md

```diff
@@ -3,49 +3,82 @@
 
 And explained how this approach should be sufficiently collision resistant:
-> Yeah, AFAIK, 20 bytes should be collision resistance when the preimages are unique and not malleable. A space of 2^160 would expect some collision to be likely around 2^80 elements (birthday paradox). And if you want to find a collision for some existing element in the database, it is still 2^160. 2^80 only is if all these elements are written to state.
 
+> Yeah, AFAIK, 20 bytes should be collision resistance when the preimages are unique and not malleable. A space of 2^160 would expect some collision to be likely around 2^80 elements (birthday paradox). And if you want to find a collision for some existing element in the database, it is still 2^160. 2^80 only is if all these elements are written to state.
 > The good example you brought up was eg. a public key bytes being a valid public key on two algorithms supported by the codec. Meaning if either was broken, you would break accounts even if they were secured with the safer variant. This is only as the issue when no differentiating type info is present in the preimage (before hashing into an address).
-
 > I would like to hear an argument if the 20 bytes space is an actual issue for security, as I would be happy to increase my address sizes in weave. I just figured cosmos and ethereum and bitcoin all use 20 bytes, it should be good enough. And the arguments above which made me feel it was secure. But I have not done a deeper analysis.
 
-In discussions in [\#5694](https://github.com/cosmos/cosmos-sdk/issues/5694), we agreed to go with an
-approach similar to this where essentially we take the first 20 bytes of the `sha256` hash of
-the key type concatenated with the key bytes, summarized as `Sha256(KeyTypePrefix || Keybytes)[:20]`.
+This lead to the first proposal (which we proved to be not good enough):
+we take the first 20 bytes of the `sha256` hash of the public key we concatenated with the key bytes, summarized as `sha256(keyTypePrefix || keybytes)[:20]`.
+
+
+### Review and Discussions
+
+In [\#5694](https://github.com/cosmos/cosmos-sdk/issues/5694) we discussed various solutions.
+We agreed that 20 bytes it's not future proof, and extending the address length is the only way to allow addresses of different types, various signature types, etc.
+This disqualifies the initial proposal.
+
+In the issue we discussed various modifications:
++ Choice of the hash function.
++ Move the prefix out of the hash function: `keyTypePrefix || sha256(keybytes)[:20]` [post-hash-prefix-proposal].
++ Use double hashing: `sha256(keyTypePrefix || sha256(keybytes)[:20])`.
++ Increase to keybytes hash slice from 20 byte to 32 or 40 bytes. We concluded that 32 bytes, produced by a good hash functions is future secure.
+
+### Requirements
+
++ Support currently used tools - we don't want to break an ecosystem, or add a long adaptation period.
```

Comment:
Can this be more explicit? What tools in particular, is the goal here to just be keeping secp key formats as is?

Reply:
Mainly wallets or services which already relay on the current addresse. More details about this is at https://github.com/cosmos/cosmos-sdk/issues/8041 I will link it directly in the ADR.